examples/cql/cql_offline.py

```diff
@@ -37,49 +42,95 @@ def main(cfg: "DictConfig"):  # noqa: F821
             experiment_name=exp_name,
             wandb_kwargs={"mode": cfg.logger.mode, "config": cfg},
         )
-
+    # Set seeds
     torch.manual_seed(cfg.env.seed)
     np.random.seed(cfg.env.seed)
     device = torch.device(cfg.optim.device)
 
-    # Make Env
+    # Create env
     train_env, eval_env = make_environment(cfg, cfg.logger.eval_envs)
 
-    # Make Buffer
+    # Create replay buffer
     replay_buffer = make_offline_replay_buffer(cfg.replay_buffer)
 
-    # Make Model
+    # Create agent
     model = make_cql_model(cfg, train_env, eval_env, device)
 
-    # Make Loss
-    loss_module, target_net_updater = make_loss(cfg.loss, model)
+    # Create loss
+    loss_module, target_net_updater = make_continuous_loss(cfg.loss, model)
 
-    # Make Optimizer
-    optimizer = make_cql_optimizer(cfg.optim, loss_module)
+    # Create Optimizer
+    (
+        policy_optim,
+        critic_optim,
+        alpha_optim,
+        alpha_prime_optim,
+    ) = make_continuous_cql_optimizer(cfg, loss_module)
 
     pbar = tqdm.tqdm(total=cfg.optim.gradient_steps)
 
-    r0 = None
-    l0 = None
-
     gradient_steps = cfg.optim.gradient_steps
+    policy_eval_start = cfg.optim.policy_eval_start
     evaluation_interval = cfg.logger.eval_iter
     eval_steps = cfg.logger.eval_steps
 
+    # Training loop
+    start_time = time.time()
     for i in range(gradient_steps):
         pbar.update(i)
+        # sample data
         data = replay_buffer.sample()
-        # loss
-        loss_vals = loss_module(data)
-        # backprop
-        actor_loss = loss_vals["loss_actor"]
+        # compute loss
+        loss_vals = loss_module(data.clone().to(device))
+
+        if i >= policy_eval_start:
+            actor_loss = loss_vals["loss_actor"]
+        else:
+            actor_loss = loss_vals["loss_actor_bc"]
```

Comment:
this seems pretty custom, can we give a small comment on why we need it?
Also, wondering how that kind of control flow would fit within a trainer class

Reply:
Indeed it's pretty custom to the CQL implementation. Just added it for completeness as it's in the official implementation. I'll add a comment on why it's there.